File: password_strength.py
"""Password strength rules and evaluation.

The rules here are the source of truth; the meter in
static/js/password-strength.js only mirrors them and is never trusted on its
own. Requirement keys match the data-req attributes in new_user.html.
"""

import re

MIN_LENGTH = 8

REQUIREMENTS = [
    ("length", f"at least {MIN_LENGTH} characters", lambda p: len(p) >= MIN_LENGTH),
    ("lowercase", "a lowercase letter", lambda p: bool(re.search(r"[a-z]", p))),
    ("uppercase", "an uppercase letter", lambda p: bool(re.search(r"[A-Z]", p))),
    ("digit", "a number", lambda p: bool(re.search(r"\d", p))),
    ("symbol", "a symbol (e.g. ! ? @ #)", lambda p: bool(re.search(r"[^A-Za-z0-9]", p))),
]

REQUIRED_COUNT = len(REQUIREMENTS)


def evaluate_password(password):
    """Summarise how a password measures up against REQUIREMENTS.

    Returns a dict with "met"/"unmet" requirement labels, "score",
    "strength" ("weak" | "fair" | "strong"), and "acceptable" (True only
    when every requirement is met).
    """
    met = [label for _key, label, test in REQUIREMENTS if test(password)]
    unmet = [label for _key, label, test in REQUIREMENTS if not test(password)]
    score = len(met)

    if score <= 2:
        strength = "weak"
    elif score < REQUIRED_COUNT:
        strength = "fair"
    else:
        strength = "strong"

    return {
        "met": met,
        "unmet": unmet,
        "score": score,
        "strength": strength,
        "acceptable": score >= REQUIRED_COUNT,
    }

File: static/js/password-strength.js
// Live password strength meter for the "Create user" form.
//
// These rules intentionally mirror password_strength.py on the server. The
// server is the real gatekeeper; this script only gives the user instant
// feedback as they type. Keep the two in sync when you change the rules.
(function () {
  "use strict";

  var MIN_LENGTH = 8;

  // Keys match the data-req attributes rendered in new_user.html.
  var RULES = [
    { key: "length", test: function (p) { return p.length >= MIN_LENGTH; } },
    { key: "lowercase", test: function (p) { return /[a-z]/.test(p); } },
    { key: "uppercase", test: function (p) { return /[A-Z]/.test(p); } },
    { key: "digit", test: function (p) { return /\d/.test(p); } },
    { key: "symbol", test: function (p) { return /[^A-Za-z0-9]/.test(p); } },
  ];

  var LABELS = ["Too weak", "Too weak", "Too weak", "Fair", "Good", "Strong"];

  var password = document.getElementById("password");
  var meter = document.querySelector("[data-strength]");
  if (!password || !meter) {
    return;
  }

  var fill = meter.querySelector("[data-strength-fill]");
  var label = meter.querySelector("[data-strength-label]");
  var items = {};
  meter.querySelectorAll("[data-req]").forEach(function (item) {
    items[item.getAttribute("data-req")] = item;
  });

  function update() {
    var value = password.value;

    if (value === "") {
      meter.hidden = true;
      return;
    }
    meter.hidden = false;

    var score = 0;
    RULES.forEach(function (rule) {
      var met = rule.test(value);
      if (met) {
        score += 1;
      }
      var item = items[rule.key];
      if (item) {
        item.classList.toggle("is-met", met);
      }
    });

    var strength = score >= RULES.length ? "strong" : score >= 3 ? "fair" : "weak";
    fill.style.width = (score / RULES.length) * 100 + "%";
    meter.setAttribute("data-level", strength);
    label.textContent = LABELS[score] || "";
  }

  password.addEventListener("input", update);
})();
